subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -0,0 +1,805 @@
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		if (obj->json) {
+			return -ENOTEMPTY;
+		}
+
+		int err;
+		cJSON *new_obj = cJSON_CreateObject();
+
+		err = (cJSON_AddStringToObjectCS(new_obj, NRF_CLOUD_JSON_APPID_KEY,
+						 app_id) ? 0 : -ENOMEM);
+
+		if (!err && msg_type) {
+			err = (cJSON_AddStringToObjectCS(new_obj,
+							 NRF_CLOUD_JSON_MSG_TYPE_KEY,
+							 msg_type) ? 0 : -ENOMEM);
+		}
+
+		if (err) {
+			cJSON_Delete(new_obj);
+			return -ENOMEM;
+		}
+
+		obj->json = new_obj;
+		return 0;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_init(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		if (obj->json) {
+			return -ENOTEMPTY;
+		}
+
+		obj->json = cJSON_CreateObject();
+		return (obj->json ? 0 : -ENOMEM);
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_reset(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		obj->json = NULL;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		return -ENOTSUP;
+	}
+
+	obj->encoded_data.len = 0;
+	obj->encoded_data.ptr = NULL;
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_bulk_init(struct nrf_cloud_obj *const bulk)
+{
+	if (!bulk) {
+		return -EINVAL;
+	}
+
+	switch (bulk->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		if (bulk->json) {
+			return -ENOTEMPTY;
+		}
+
+		bulk->json = cJSON_CreateArray();
+		return (bulk->json ? 0 : -ENOMEM);
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_cloud_encoded_free(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	if (obj->enc_src != NRF_CLOUD_ENC_SRC_CLOUD_ENCODED) {
+		return -EACCES;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		if (obj->encoded_data.ptr) {
+			cJSON_free((void *)obj->encoded_data.ptr);
+			obj->encoded_data.ptr = NULL;
+		}
+
+		obj->encoded_data.len = 0;
+		obj->enc_src = NRF_CLOUD_ENC_SRC_NONE;
+		return 0;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_cloud_pre_encoded_free(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	if (obj->enc_src != NRF_CLOUD_ENC_SRC_PRE_ENCODED) {
+		return -EACCES;
+	}
+
+	if (obj->encoded_data.ptr) {
+		nrf_cloud_free((void *)obj->encoded_data.ptr);
+		obj->encoded_data.ptr = NULL;
+	}
+
+	obj->encoded_data.len = 0;
+	obj->enc_src = NRF_CLOUD_ENC_SRC_NONE;
+
+	return 0;
+}
+
+
+int nrf_cloud_obj_free(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		cJSON_Delete(obj->json);
+		obj->json = NULL;
+		return 0;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
```

Comment:
`nrf_cloud_obj_cloud_encoded_free` was added just to be complete... maybe it doesn't make sense.  i used `nrf_cloud_free` so the proper free function would be used.

i feel that `nrf_cloud_obj_free` and `nrf_cloud_obj_cloud_encoded_free` should be separate functions as the memory is allocated by separate functions.  this gives the control to the user to free the specific item when they want.

Reply:
> i feel that nrf_cloud_obj_free and nrf_cloud_obj_cloud_encoded_free should be separate functions as the memory is allocated by separate functions. this gives the control to the user to free the specific item when they want.

That makes sense, as long as we really do want to allow users to free just the encoded contents without freeing the whole message

But do we really want that? I think there is an opportunity for cleanliness if we say "no, user, don't worry about the encoded data, just worry about freeing `nrf_cloud_obj` and we will keep track of the underlying encoded data for you"

Maybe I am wrong, though, and we really should offer that ability to the end user

Reply:
Ah, thinking on it, your argument also makes more sense if we're talking about a stack-allocated `nrf_cloud_obj` as opposed to one that is heap-allocated.

Otherwise user might, in my suggestion, call `nrf_cloud_obj_cloud_encode` on a stack-allocated `nrf_cloud_obj` and it won't be nearly as obvious that doing so means they must then call `nrf_cloud_free` on the object. I catch your drift now.

I wonder, then, though: Does the user need access to `nrf_cloud_obj_cloud_encode`, or can this be something internal to the nrf_cloud lib? Privatizing it might make this lifetime issue simpler to design around

Reply:
currently `nrf_cloud_obj_cloud_encode` is only used internally by the nrf_cloud library.
so, yes it could be removed from the public interface.
however, i would lean towards leaving the capability publicly available in case there is some use-case not covered by the samples i've updated.  i think as long as it is documented properly, then it can stay public? 

i will remove `nrf_cloud_obj_cloud_pre_encoded_free.`.